cpukit/score/src/schedulerstrongapa.c

```diff
@@ -3,339 +3,820 @@
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
  */
 
 #ifdef HAVE_CONFIG_H
 #include "config.h"
 #endif
 
 #include <rtems/score/schedulerstrongapa.h>
-#include <rtems/score/schedulerpriorityimpl.h>
 #include <rtems/score/schedulersmpimpl.h>
+#include <rtems/score/assert.h>
 
-static Scheduler_strong_APA_Context *_Scheduler_strong_APA_Get_self(
-  Scheduler_Context *context
-)
+static inline Scheduler_strong_APA_Context *
+_Scheduler_strong_APA_Get_context( const Scheduler_Control *scheduler )
+{
+  return (Scheduler_strong_APA_Context *) _Scheduler_Get_context( scheduler );
+}
+
+static inline Scheduler_strong_APA_Context *
+_Scheduler_strong_APA_Get_self( Scheduler_Context *context )
 {
   return (Scheduler_strong_APA_Context *) context;
 }
 
-static Scheduler_strong_APA_Node *
+static inline Scheduler_strong_APA_Node *
 _Scheduler_strong_APA_Node_downcast( Scheduler_Node *node )
 {
   return (Scheduler_strong_APA_Node *) node;
 }
 
-static void _Scheduler_strong_APA_Move_from_scheduled_to_ready(
+static inline void _Scheduler_strong_APA_Do_update(
   Scheduler_Context *context,
-  Scheduler_Node    *scheduled_to_ready
+  Scheduler_Node    *node,
+  Priority_Control   new_priority
 )
 {
-  Scheduler_strong_APA_Context *self =
-    _Scheduler_strong_APA_Get_self( context );
-  Scheduler_strong_APA_Node *node =
-    _Scheduler_strong_APA_Node_downcast( scheduled_to_ready );
-
-  _Chain_Extract_unprotected( &node->Base.Base.Node.Chain );
-  _Scheduler_priority_Ready_queue_enqueue_first(
-    &node->Base.Base.Node.Chain,
-    &node->Ready_queue,
-    &self->Bit_map
-  );
+  Scheduler_SMP_Node *smp_node;
+  (void) context;
+
+  smp_node = _Scheduler_SMP_Node_downcast( node );
+  _Scheduler_SMP_Node_update_priority( smp_node, new_priority );
+}
+
+/*
+ * Returns true if the Strong APA scheduler has ready nodes
+ * available for scheduling.
+ */
+static inline bool
+    _Scheduler_strong_APA_Has_ready( Scheduler_Context *context )
+{
+  Scheduler_strong_APA_Context *self;
+  const Chain_Node             *tail;
+  Chain_Node                   *next;
+  Scheduler_strong_APA_Node    *node;
+
+  self = _Scheduler_strong_APA_Get_self( context );
+  tail = _Chain_Immutable_tail( &self->Ready );
+  next = _Chain_First( &self->Ready );
+
+  while ( next != tail ) {
+    node = (Scheduler_strong_APA_Node *) STRONG_SCHEDULER_NODE_OF_CHAIN( next );
+
+    if (
+      _Scheduler_SMP_Node_state( &node->Base.Base ) ==
+      SCHEDULER_SMP_NODE_READY
+    ) {
```

Comment:
Please explain what you mean by timing inconsistency. If the context stores a bool variable has_ready, I think every time a node is extracted from Ready_queue, we would need to update the variable - by traversing the queue :p! Do you have a better implementation in mind? Please note that Ready_queue stores both ready and scheduled nodes, so we won't be able to simply check if _Chain_first->next == _Chain_tail.

Reply:
The timing issue is whether the list can be accessed in parallel to this operation?

You could keep a count of ready nodes?

Reply:
Good point. Two things: 1)The only function that calls the _Scheduler_strong_APA_Has_ready is _Scheduler_strong_APA_Add_processor, and as far as I remember the latter function is called when we explicitly add a processor during the runtime on an application right? (Because I set a breakpoint at _Scheduler_strong_APA_Add_processor, but the program did not stop). So should we worry so much about the runtime of this, if it is so rarely going to be used?
2) About the timing issue: Does RTEMS not acquire some kind of lock in SMP systems every time we access a scheduler function from the application - so that other cores also accessing any scheduler function do not cause unexpected behaviour ? I can see :_Thread_State_acquire_critical( the_thread, lock_context );: gets used a lot - what does it lock?

Reply:
"I set a breakpoint at _Scheduler_strong_APA_Add_processor, but the program did not stop" -- if no test hits this, that is a different problem. You should attempt to design a test to hit this.

Reply:
"Does RTEMS not acquire some kind of lock in SMP systems every time we access a scheduler function from the application - so that other cores also accessing any scheduler function do not cause unexpected behaviour" -- Maybe you should dig into this as well. I don't think your assumption is correct. There are multiple kinds of locks.

Reply:
" - "I set a breakpoint at _Scheduler_strong_APA_Add_processor, but the program did not stop" -- if no test hits this, that is a different problem. You should attempt to design a test to hit this. -"
I only ran the smp01 test to check this, and it did not hit this. I will ask on devel if there's a test that tests this.

Reply:
Okay, so with your help I know that smpscheduler04 uses add_processor directive, and still passes. Hence this function works. We also know that it is not subject to a timing inconsistency because any function that accesses the chain in Strong APA's context has to acquire a scheduler lock first, which only one core/function can get at any particular time. 

So, is it okay if we think about improving the run-time of this function later, maybe in the next version?

Reply:
yes

Reply:
Thanks!